### ISSUE TITLE
transition-based acceptance marks in alternating automata

### ISSUE THREAD

**Issue by adl** [adl/hoaf#68]
I wonder if it was a mistake to consider that a transition with universal branching can have a single set of acceptance marks: I believe it would make more sense to have one such set per destination of universal transition.

Let me justify this statement by considering the *dualization* of an automaton, an operation that will be in the next major release of Spot.

A complete alternating automaton with state-based acceptance, can be complemented by simply complementing the acceptance condition, and dualizing the transition structures (i.e. existential choices become universal and vice-versa).

```
ltl2tgba -C FGa
```
![in](https://user-images.githubusercontent.com/822900/27949165-d80bf34e-62fc-11e7-8191-a47033a5735f.png)
```
ltl2tgba -C FGa | autfilt --dualize
```
![out](https://user-images.githubusercontent.com/822900/27949222-2e3b0fca-62fd-11e7-8486-d1701044949d.png)

The transformation is quite easy to implement, and works for any acceptance.  It does not add new states, and the total number of destinations of transitions does not change.

Now let's transpose this algorithm to alternating automata with transition-based acceptance.
For instance here is a handcrafted automaton recognizing `GFa`:

![int](https://user-images.githubusercontent.com/822900/27956017-05c538d4-6318-11e7-8d57-8e0722e9b2f6.png)

Dualizing the above automaton as follows would be clearly incorrect, since it would accept any word.
![outt](https://user-images.githubusercontent.com/822900/27956058-41d74286-6318-11e7-82d9-7febc165a3a5.png)

What I'd really want is the following, that HOA cannot represent:
![outt](https://user-images.githubusercontent.com/822900/27956294-6d79a5f4-6319-11e7-8383-5cd841ab6825.png)

Note that in this case, pushing the acceptance marks to states 1 and 2 is possible without adding more states, so we could use the state-based dualization.  However we can render this case more complex by making sure that states 1 or 2 are destination of other transitions with different marks.

**Comment by strejcek** [comment]
I don't think that it was a mistake to consider that a transition with universal branching can have a single set of acceptance marks. There just wasn't any need to refine this notion at that time. By the way, the only result combining alternating automata with transition-based acceptance I know about is the unpublished ltl3hoa transition (and it works just with acceptance sets assigned to whole transitions). 

Anyway, I see your point with dualization. I like the idea, but I'm not sure whether it pays to refining the transition-based acceptance in the format  as it is a lot of work and there are not many applications. But I'm not against: if you want it, suggest changes and I'll carefully review and polish your pull requests.

**Comment by adl** [comment]
> the only result combining alternating automata with transition-based acceptance I know about is the unpublished ltl3hoa transition

The first published result I know is from 2006.  [Heikki Tauriainen's Ph.D. thesis](http://lib.tkk.fi/Diss/2006/isbn9512283433/isbn9512283433.pdf) introduced alternating automata with a set of acceptance marks per transition (bottom of page 13), and proposed a translation from LTL.  Heikki never implemented his translation, but I had a student implement it in Spot 1.0 in 2010 using some ad-hoc data structures for alternating automata.  Today that translation is still present in Spot, but it hasn't been ported to the "official" alternating-automata data structures and therefore we still cannot output it in HOA.  `ltl3hoa` is at least the first tool to output transition-based alternating automata in HOA.

BTW, Heikki's thesis is also the first document were I saw acceptance sets represented by small bullets on transitions; an idea that I've reused since then.

>  I see your point with dualization. I like the idea, but I'm not sure whether it pays to refine the transition-based acceptance in the format: it is a lot of work everywhere and there are not many applications.

Changing the data structures in Spot to support such a change would be a lot of work and backward-incompatible changes, so I'm not eager to update the format either.  This issue is mostly a way to share a disappointment: our formalism is not as elegant as it could be since dualization of transition-based acceptance cannot be done naturally.  But so far, it's an issue for only one algorithm; let's wait and see if we find more.

**Comment by strejcek** [comment]
> The first published result I know is from 2006. Heikki Tauriainen's Ph.D. thesis introduced alternating automata with a set of acceptance marks per transition (bottom of page 13) [...]

Oh yes, I've seen the thesis a long time ago... 

> This issue is mostly a way to share a disappointment: our formalism is not as elegant as it could be since dualization of transition-based acceptance cannot be done naturally. But so far, it's an issue for only one algorithm; let's wait and see if we find more.

I fully agree. Lets keep it like this for now.
